common/encoding.h

```diff
@@ -0,0 +1,241 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#ifndef COMMON_ENCODING_H
+#define COMMON_ENCODING_H
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif // HAVE_CONFIG_H
+
+#ifdef USE_ICONV
+#include <iconv.h>
+#else
+typedef void* iconv_t;
+#endif // USE_ICONV
+
+#include "common/scummsys.h"
+#include "common/str.h"
+
+namespace Common {
+
+/**
+ * A class, that allows conversion between different text encoding,
+ * the encodings available depend on the current backend and if the
+ * ScummVM is compiled with or without iconv.
+ */
+class Encoding {
+	public:
+		/**
+		 * Constructs everything needed for the conversion between 2 encodings
+		 * and saves the values for future use.
+		 *
+		 * @param to Name of the encoding the strings will be converted to
+		 * @param from Name of the encoding the strings will be converted from
+		 */
+		Encoding(const String &to, const String &from);
+		~Encoding();
+
+		/**
+		 * Converts string between encodings. The resulting string is ended by 
+		 * a character with value 0 (C-like ending for 1 byte per character
+		 * encodings, 2 zero bytes for UTF-16, 4 zero bytes for UTF-32)
+		 *
+		 * The result has to be freed after use.
+		 *
+		 * @param string String that should be converted.
+		 * @param length Length of the string to convert in bytes.
+		 *
+		 * @return Converted string (must be freed) or nullptr if the conversion failed
+		 */
+		char *convert(const char *string, size_t length);
+
+		/**
+		 * Static version of the method above.
+		 * Converts string between encodings. The resulting string is ended by 
+		 * a character with value 0 (C-like ending for 1 byte per character
+		 * encodings, 2 zero bytes for UTF-16, 4 zero bytes for UTF-32)
+		 *
+		 * The result has to be freed after use.
```

Comment:
Why? Return a Common::String instead of a char *

Reply:
This should be fine for most encodings. But I thought using Common::String doesn't really make sense for multi byte encodings (UTF-16, UTF-32), because Common::String doesn't seem like it is made for multibyte encodings (Although it should be fine to use it just for returning from the Common::Encoding) So I thought it would be better to return an array of chars and let the person who uses it to decide what to do with it.

Reply:
True. We have a class for UTF32 strings, U32String, in common/ustr.*

You should use that here

Reply:
Yes, but the U32String is made for UTF-32. What should I do with the rest of the encodings? I might be able to somehow return UTF-32 results as Common::U32String and one byte per character results as Common::String. But what about the UTF-16?